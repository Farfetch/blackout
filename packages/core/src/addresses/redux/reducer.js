/**
 * @module addresses/reducer
 * @category Addresses
 * @subcategory Reducer
 */

import * as actionTypes from './actionTypes';
import { combineReducers } from 'redux';
import { createReducerWithResult, reducerFactory } from '../../helpers/redux';
import get from 'lodash/get';
import omit from 'lodash/omit';
import produce from 'immer';

const INITIAL_STATE = {
  error: null,
  isLoading: false,
  result: null,
  predictions: {
    result: null,
    error: null,
    isLoading: false,
  },
  predictionDetails: {
    result: null,
    error: null,
    isLoading: false,
  },
  addresses: {
    error: null,
    isLoading: false,
  },
  address: {
    error: {},
    isLoading: {},
  },
  addressSchema: {
    error: null,
    isLoading: false,
  },
  /* Used for operations related with the default address that
    have a result associated, such as getting the default contact address */
  defaultAddressDetails: {
    error: null,
    isLoading: false,
    result: null,
  },
};

export const getDefaultAddress = (addressesList, prop) => {
  for (const key in addressesList) {
    if (addressesList[key] && addressesList[key][prop]) {
      return addressesList[key];
    }
  }
  return null;
};

const result = (state = INITIAL_STATE.result, action = {}) => {
  switch (action.type) {
    case actionTypes.GET_ADDRESSES_SUCCESS:
      return action.payload.result;
    case actionTypes.CREATE_ADDRESS_SUCCESS:
      return [...state, action.meta.addressId];
    case actionTypes.DELETE_ADDRESS_SUCCESS: {
      return (
        state && state.filter(addressId => addressId !== action.meta.addressId)
      );
    }
    default:
      return state;
  }
};

const error = (state = INITIAL_STATE.error, action = {}) => {
  switch (action.type) {
    case actionTypes.GET_PREDICTION_FAILURE:
    case actionTypes.GET_PREDICTION_DETAILS_FAILURE:
    case actionTypes.GET_ADDRESS_FAILURE:
    case actionTypes.GET_ADDRESSES_FAILURE:
    case actionTypes.CREATE_ADDRESS_FAILURE:
    case actionTypes.UPDATE_ADDRESS_FAILURE:
    case actionTypes.SET_DEFAULT_BILLING_ADDRESS_FAILURE:
    case actionTypes.SET_DEFAULT_SHIPPING_ADDRESS_FAILURE:
    case actionTypes.SET_DEFAULT_CONTACT_ADDRESS_FAILURE:
    case actionTypes.DELETE_DEFAULT_CONTACT_ADDRESS_FAILURE:
    case actionTypes.DELETE_DEFAULT_SHIPPING_ADDRESS_FAILURE:
    case actionTypes.DELETE_DEFAULT_BILLING_ADDRESS_FAILURE:
    case actionTypes.GET_DEFAULT_CONTACT_ADDRESS_FAILURE:
      return action.payload.error;
    case actionTypes.GET_PREDICTION_REQUEST:
    case actionTypes.GET_PREDICTION_DETAILS_REQUEST:
    case actionTypes.GET_ADDRESS_REQUEST:
    case actionTypes.GET_ADDRESSES_REQUEST:
    case actionTypes.CREATE_ADDRESS_REQUEST:
    case actionTypes.UPDATE_ADDRESS_REQUEST:
    case actionTypes.SET_DEFAULT_BILLING_ADDRESS_REQUEST:
    case actionTypes.SET_DEFAULT_SHIPPING_ADDRESS_REQUEST:
    case actionTypes.SET_DEFAULT_CONTACT_ADDRESS_REQUEST:
    case actionTypes.DELETE_DEFAULT_CONTACT_ADDRESS_REQUEST:
    case actionTypes.GET_DEFAULT_CONTACT_ADDRESS_REQUEST:
    case actionTypes.DELETE_DEFAULT_BILLING_ADDRESS_REQUEST:
    case actionTypes.DELETE_DEFAULT_SHIPPING_ADDRESS_REQUEST:
      return INITIAL_STATE.error;
    default:
      return state;
  }
};

const isLoading = (state = INITIAL_STATE.isLoading, action = {}) => {
  switch (action.type) {
    case actionTypes.GET_PREDICTION_REQUEST:
    case actionTypes.GET_PREDICTION_DETAILS_REQUEST:
    case actionTypes.GET_ADDRESSES_REQUEST:
    case actionTypes.GET_ADDRESS_REQUEST:
    case actionTypes.CREATE_ADDRESS_REQUEST:
    case actionTypes.UPDATE_ADDRESS_REQUEST:
    case actionTypes.SET_DEFAULT_BILLING_ADDRESS_REQUEST:
    case actionTypes.SET_DEFAULT_SHIPPING_ADDRESS_REQUEST:
    case actionTypes.SET_DEFAULT_CONTACT_ADDRESS_REQUEST:
    case actionTypes.DELETE_DEFAULT_CONTACT_ADDRESS_REQUEST:
    case actionTypes.DELETE_DEFAULT_BILLING_ADDRESS_REQUEST:
    case actionTypes.DELETE_DEFAULT_SHIPPING_ADDRESS_REQUEST:
    case actionTypes.GET_DEFAULT_CONTACT_ADDRESS_REQUEST:
      return true;
    case actionTypes.GET_PREDICTION_FAILURE:
    case actionTypes.GET_PREDICTION_SUCCESS:
    case actionTypes.GET_PREDICTION_DETAILS_FAILURE:
    case actionTypes.GET_PREDICTION_DETAILS_SUCCESS:
    case actionTypes.GET_ADDRESSES_FAILURE:
    case actionTypes.GET_ADDRESSES_SUCCESS:
    case actionTypes.GET_ADDRESS_FAILURE:
    case actionTypes.GET_ADDRESS_SUCCESS:
    case actionTypes.CREATE_ADDRESS_FAILURE:
    case actionTypes.CREATE_ADDRESS_SUCCESS:
    case actionTypes.UPDATE_ADDRESS_FAILURE:
    case actionTypes.UPDATE_ADDRESS_SUCCESS:
    case actionTypes.SET_DEFAULT_BILLING_ADDRESS_FAILURE:
    case actionTypes.SET_DEFAULT_BILLING_ADDRESS_SUCCESS:
    case actionTypes.SET_DEFAULT_SHIPPING_ADDRESS_FAILURE:
    case actionTypes.SET_DEFAULT_SHIPPING_ADDRESS_SUCCESS:
    case actionTypes.SET_DEFAULT_CONTACT_ADDRESS_FAILURE:
    case actionTypes.SET_DEFAULT_CONTACT_ADDRESS_SUCCESS:
    case actionTypes.DELETE_DEFAULT_CONTACT_ADDRESS_FAILURE:
    case actionTypes.DELETE_DEFAULT_CONTACT_ADDRESS_SUCCESS:
    case actionTypes.DELETE_DEFAULT_BILLING_ADDRESS_FAILURE:
    case actionTypes.DELETE_DEFAULT_BILLING_ADDRESS_SUCCESS:
    case actionTypes.DELETE_DEFAULT_SHIPPING_ADDRESS_FAILURE:
    case actionTypes.DELETE_DEFAULT_SHIPPING_ADDRESS_SUCCESS:
    case actionTypes.GET_DEFAULT_CONTACT_ADDRESS_FAILURE:
    case actionTypes.GET_DEFAULT_CONTACT_ADDRESS_SUCCESS:
      return INITIAL_STATE.isLoading;
    default:
      return state;
  }
};

export const predictions = (state = INITIAL_STATE.predictions, action = {}) => {
  switch (action.type) {
    case actionTypes.GET_PREDICTION_REQUEST:
      return {
        error: INITIAL_STATE.predictions.error,
        isLoading: true,
      };
    case actionTypes.GET_PREDICTION_FAILURE:
      return {
        error: action.payload.error,
        isLoading: false,
      };
    case actionTypes.GET_PREDICTION_SUCCESS:
      return {
        error: INITIAL_STATE.predictions.error,
        isLoading: false,
        result: action.payload,
      };
    case actionTypes.RESET_PREDICTION:
      return INITIAL_STATE.predictions;
    default:
      return state;
  }
};

export const predictionDetails = createReducerWithResult(
  'GET_PREDICTION_DETAILS',
  INITIAL_STATE.predictionDetails,
  actionTypes,
);

export const entitiesMapper = {
  [actionTypes.CREATE_ADDRESS_SUCCESS]: (state, action) => {
    const id = action.payload.result;
    const createdAddress = action.payload.entities.addresses[id];

    return produce(state, draftState => {
      if (!draftState.addresses) {
        draftState.addresses = {};
      }

      draftState.addresses = {
        ...draftState.addresses,
        [id]: { ...createdAddress },
      };
    });
  },
  [actionTypes.UPDATE_ADDRESS_SUCCESS]: (state, action) => {
    const id = action.payload.result;
    const updatedAddress = action.payload.entities.addresses[id];

    return produce(state, draftState => {
      if (!draftState.addresses) {
        draftState.addresses = {};
      }

      draftState.addresses = {
        ...draftState.addresses,
        [id]: { ...updatedAddress },
      };
    });
  },
  [actionTypes.DELETE_ADDRESS_SUCCESS]: (state, action) => {
    const { addressId } = action.meta;
    const currentAddresses = state.addresses;

    return produce(state, draftState => {
      draftState.addresses = omit(currentAddresses, addressId);
    });
  },
  [actionTypes.SET_DEFAULT_SHIPPING_ADDRESS_SUCCESS]: (state, action) => {
    const { addressId } = action.meta;

    // Get prev default address so it can later be unmarked as the default
    const prevDefaultShippingAddress = getDefaultAddress(
      state.addresses,
      'isDefaultShippingAddress',
    );

    return produce(state, draftState => {
      if (prevDefaultShippingAddress) {
        // Unmark previous shipping address as default
        draftState.addresses[
          prevDefaultShippingAddress.id
        ].isDefaultShippingAddress = false;
      }
      // Select the selected address as default
      draftState.addresses[addressId].isDefaultShippingAddress = true;
    });
  },
  [actionTypes.SET_DEFAULT_BILLING_ADDRESS_SUCCESS]: (state, action) => {
    const { addressId } = action.meta;

    // Get prev default address so it can later be unmarked as the default
    const prevDefaultBillingAddress = getDefaultAddress(
      state.addresses,
      'isDefaultBillingAddress',
    );

    return produce(state, draftState => {
      if (prevDefaultBillingAddress) {
        // Unmark previous billing address as default
        draftState.addresses[
          prevDefaultBillingAddress.id
        ].isDefaultBillingAddress = false;
      }
      // Select the selected address as default
      draftState.addresses[addressId].isDefaultBillingAddress = true;
    });
  },
  [actionTypes.GET_ADDRESS_SCHEMA_SUCCESS]: (state, action) => {
    const countryId = action.payload.result;
    const countrySchema = get(
      action,
      `payload.entities.addressSchema[${countryId}]`,
      {},
    );

    return produce(state, draftState => {
      if (!draftState.addressSchema) {
        draftState.addressSchema = {};
      }

      draftState.addressSchema[countryId] = { ...countrySchema };
    });
  },
  [actionTypes.SET_DEFAULT_CONTACT_ADDRESS_SUCCESS]: (state, action) => {
    const { addressId } = action.meta;

    // Get prev default address so it can later be unmarked as the default
    const prevDefaultContactAddress = getDefaultAddress(
      state.addresses,
      'isPreferredAddress',
    );

    return produce(state, draftState => {
      if (prevDefaultContactAddress) {
        // Unmark previous contact address as default
        draftState.addresses[
          prevDefaultContactAddress.id
        ].isPreferredAddress = false;
      }
      // Select the selected address as default
      draftState.addresses[addressId].isPreferredAddress = true;
    });
  },
  [actionTypes.DELETE_DEFAULT_CONTACT_ADDRESS_SUCCESS]: (state, action) => {
    const { addressId } = action.meta;

    return produce(state, draftState => {
      // Unmark the selected address as default
      draftState.addresses[addressId].isPreferredAddress = false;
    });
  },
  [actionTypes.DELETE_DEFAULT_BILLING_ADDRESS_SUCCESS]: (state, action) => {
    const { addressId } = action.meta;

    return produce(state, draftState => {
      if (!draftState) {
        return draftState;
      }

      const addresses = draftState.addresses;

      if (!addresses) {
        return draftState;
      }

      // Unmark the selected address as default
      const defaultBillingAddress = addresses[addressId];

      if (defaultBillingAddress) {
        defaultBillingAddress.isCurrentBilling = false;
      }

      return draftState;
    });
  },
  [actionTypes.DELETE_DEFAULT_SHIPPING_ADDRESS_SUCCESS]: (state, action) => {
    const { addressId } = action.meta;

    return produce(state, draftState => {
      if (!draftState) {
        return draftState;
      }

      const addresses = draftState.addresses;

      if (!addresses) {
        return draftState;
      }

      // Unmark the selected address as default
      const defaultShippingAddress = addresses[addressId];

      if (defaultShippingAddress) {
        defaultShippingAddress.isCurrentShipping = false;
      }

      return draftState;
    });
  },
};

export const address = (state = INITIAL_STATE.address, action = {}) => {
  switch (action.type) {
    case actionTypes.CREATE_ADDRESS_REQUEST:
      return {
        isLoading: { ...state.isLoading },
        error: {},
      };
    case actionTypes.GET_ADDRESS_REQUEST:
    case actionTypes.UPDATE_ADDRESS_REQUEST:
    case actionTypes.DELETE_ADDRESS_REQUEST:
    case actionTypes.SET_DEFAULT_BILLING_ADDRESS_REQUEST:
    case actionTypes.SET_DEFAULT_SHIPPING_ADDRESS_REQUEST:
    case actionTypes.SET_DEFAULT_CONTACT_ADDRESS_REQUEST:
    case actionTypes.DELETE_DEFAULT_CONTACT_ADDRESS_REQUEST:
    case actionTypes.DELETE_DEFAULT_BILLING_ADDRESS_REQUEST:
    case actionTypes.DELETE_DEFAULT_SHIPPING_ADDRESS_REQUEST:
      return {
        isLoading: {
          ...state.isLoading,
          [action.meta.addressId]: true,
        },
        error: {},
      };
    case actionTypes.GET_ADDRESS_SUCCESS:
    case actionTypes.UPDATE_ADDRESS_SUCCESS:
    case actionTypes.DELETE_ADDRESS_SUCCESS:
    case actionTypes.SET_DEFAULT_BILLING_ADDRESS_SUCCESS:
    case actionTypes.SET_DEFAULT_SHIPPING_ADDRESS_SUCCESS:
    case actionTypes.SET_DEFAULT_CONTACT_ADDRESS_SUCCESS:
    case actionTypes.DELETE_DEFAULT_CONTACT_ADDRESS_SUCCESS:
    case actionTypes.DELETE_DEFAULT_BILLING_ADDRESS_SUCCESS:
    case actionTypes.DELETE_DEFAULT_SHIPPING_ADDRESS_SUCCESS:
      return {
        ...state,
        isLoading: {
          ...state.isLoading,
          [action.meta.addressId]: false,
        },
      };
    case actionTypes.GET_ADDRESS_FAILURE:
    case actionTypes.UPDATE_ADDRESS_FAILURE:
    case actionTypes.DELETE_ADDRESS_FAILURE:
    case actionTypes.SET_DEFAULT_BILLING_ADDRESS_FAILURE:
    case actionTypes.SET_DEFAULT_SHIPPING_ADDRESS_FAILURE:
    case actionTypes.SET_DEFAULT_CONTACT_ADDRESS_FAILURE:
    case actionTypes.DELETE_DEFAULT_CONTACT_ADDRESS_FAILURE:
    case actionTypes.DELETE_DEFAULT_BILLING_ADDRESS_FAILURE:
    case actionTypes.DELETE_DEFAULT_SHIPPING_ADDRESS_FAILURE:
      return {
        ...state,
        isLoading: {
          ...state.isLoading,
          [action.meta.addressId]: false,
        },
        error: {
          ...state.error,
          [action.meta.addressId]: action.payload.error,
        },
      };
    default:
      return state;
  }
};

export const addresses = reducerFactory(
  'GET_ADDRESSES',
  INITIAL_STATE.addresses,
  actionTypes,
);

export const addressSchema = reducerFactory(
  'GET_ADDRESS_SCHEMA',
  INITIAL_STATE.addressSchema,
  actionTypes,
);

export const defaultAddressDetails = createReducerWithResult(
  'GET_DEFAULT_CONTACT_ADDRESS',
  INITIAL_STATE.defaultAddressDetails,
  actionTypes,
);

export const getError = state => state.error;
export const getIsLoading = state => state.isLoading;
export const getResult = state => state.result;

export const getPredictions = state => state.predictions;
export const getPredictionDetails = state => state.predictionDetails;
export const getAddresses = state => state.addresses;
export const getAddress = state => state.address;
export const getAddressSchema = state => state.addressSchema;
export const getDefaultAddressDetails = state => state.defaultAddressDetails;

const reducer = combineReducers({
  error,
  isLoading,
  result,
  predictions,
  predictionDetails,
  addresses,
  address,
  addressSchema,
  defaultAddressDetails,
});

/**
 * Reducer for addresses state.
 *
 * @function addressesReducer
 * @static
 *
 * @param {object} state - Current redux state.
 * @param {object} action - Action dispatched.
 *
 * @returns {object} New state.
 */
export default (state, action = {}) => {
  if (action.type === actionTypes.RESET_ADDRESSES) {
    return INITIAL_STATE;
  }

  return reducer(state, action);
};
