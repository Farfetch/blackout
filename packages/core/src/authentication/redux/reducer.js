/**
 * @module authentication/reducer
 * @category Authentication
 * @subcategory Reducer
 */

import * as actionTypes from './actionTypes';
import { combineReducers } from 'redux';
import { reducerFactory } from '../../helpers/redux';

const INITIAL_STATE = {
  error: null,
  id: null,
  isLoading: false,
  login: {
    error: null,
    isLoading: false,
  },
  logout: {
    error: null,
    isLoading: false,
  },
  register: {
    error: null,
    isLoading: false,
  },
  changePassword: {
    error: null,
    isLoading: false,
  },
  recoverPassword: {
    error: null,
    isLoading: false,
  },
  resetPassword: {
    error: null,
    isLoading: false,
  },
  validateEmail: {
    error: null,
    isLoading: false,
  },
  // TODO: Remove in version 2.0
  refreshToken: {
    error: null,
    isLoading: false,
  },
  refreshEmailToken: {
    error: null,
    isLoading: false,
  },
};

const error = (state = INITIAL_STATE.error, action = {}) => {
  switch (action.type) {
    case actionTypes.LOGIN_FAILURE:
    case actionTypes.LOGOUT_FAILURE:
    case actionTypes.REGISTER_FAILURE:
    case actionTypes.PASSWORD_CHANGE_FAILURE:
    case actionTypes.PASSWORD_RECOVER_FAILURE:
    case actionTypes.PASSWORD_RESET_FAILURE:
      return action.payload.error;
    case actionTypes.AUTHENTICATION_RESET:
    case actionTypes.LOGIN_REQUEST:
    case actionTypes.LOGOUT_REQUEST:
    case actionTypes.PASSWORD_CHANGE_REQUEST:
    case actionTypes.PASSWORD_RECOVER_REQUEST:
    case actionTypes.PASSWORD_RESET_REQUEST:
    case actionTypes.REGISTER_REQUEST:
      return INITIAL_STATE.error;
    default:
      return state;
  }
};

const id = (state = INITIAL_STATE.id, action = {}) => {
  switch (action.type) {
    case actionTypes.LOGIN_SUCCESS:
    case actionTypes.REGISTER_SUCCESS:
      return action.payload.result;
    case actionTypes.LOGOUT_SUCCESS:
    case actionTypes.AUTHENTICATION_RESET:
      return INITIAL_STATE.id;
    default:
      return state;
  }
};

const isLoading = (state = INITIAL_STATE.isLoading, action = {}) => {
  switch (action.type) {
    case actionTypes.LOGIN_REQUEST:
    case actionTypes.LOGOUT_REQUEST:
    case actionTypes.REGISTER_REQUEST:
    case actionTypes.PASSWORD_CHANGE_REQUEST:
    case actionTypes.PASSWORD_RECOVER_REQUEST:
    case actionTypes.PASSWORD_RESET_REQUEST:
      return true;
    case actionTypes.LOGIN_SUCCESS:
    case actionTypes.LOGOUT_SUCCESS:
    case actionTypes.REGISTER_SUCCESS:
    case actionTypes.PASSWORD_CHANGE_SUCCESS:
    case actionTypes.PASSWORD_RECOVER_SUCCESS:
    case actionTypes.PASSWORD_RESET_SUCCESS:
      return INITIAL_STATE.isLoading;
    default:
      return state;
  }
};

export const login = reducerFactory('LOGIN', INITIAL_STATE.login, actionTypes);
export const logout = reducerFactory(
  'LOGOUT',
  INITIAL_STATE.logout,
  actionTypes,
);
export const register = reducerFactory(
  'REGISTER',
  INITIAL_STATE.register,
  actionTypes,
);
export const changePassword = reducerFactory(
  'PASSWORD_CHANGE',
  INITIAL_STATE.changePassword,
  actionTypes,
);
export const resetPassword = reducerFactory(
  'PASSWORD_RESET',
  INITIAL_STATE.resetPassword,
  actionTypes,
);
export const recoverPassword = reducerFactory(
  'PASSWORD_RECOVER',
  INITIAL_STATE.recoverPassword,
  actionTypes,
);
export const validateEmail = reducerFactory(
  'VALIDATE_EMAIL',
  INITIAL_STATE.validateEmail,
  actionTypes,
);
export const refreshToken = reducerFactory(
  'REFRESH_TOKEN',
  INITIAL_STATE.refreshToken,
  actionTypes,
);
export const refreshEmailToken = reducerFactory(
  'REFRESH_EMAIL_TOKEN',
  INITIAL_STATE.refreshEmailToken,
  actionTypes,
);

export const entitiesMapper = {
  [actionTypes.LOGOUT_SUCCESS]: state => {
    return {
      ...state,
      user: {},
    };
  },
};

export const getError = state => state.error;
export const getIsLoading = state => state.isLoading;
export const getId = state => state.id;
export const getLogin = state => state.login;
export const getLogout = state => state.logout;
export const getRegister = state => state.register;
export const getChangePassword = state => state.changePassword;
export const getResetPassword = state => state.resetPassword;
export const getRecoverPassword = state => state.recoverPassword;
export const getValidateEmail = state => state.validateEmail;
export const getRefreshEmailToken = state => state.refreshEmailToken;
export const getRefreshToken = state => state.refreshToken; // @TODO: Remove in version 2.0

/**
 * Reducer for authentication state.
 *
 * @function authenticationReducer
 * @static
 *
 * @param {object} state - Current redux state.
 * @param {object} action - Action dispatched.
 *
 * @returns {object} New state.
 */
export default combineReducers({
  error,
  id,
  isLoading,
  login,
  logout,
  register,
  changePassword,
  resetPassword,
  recoverPassword,
  validateEmail,
  refreshToken, // @TODO: To remove in the next version 2.0
  refreshEmailToken,
});
