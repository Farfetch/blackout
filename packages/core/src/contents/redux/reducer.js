/**
 * @module contents/reducer
 * @category Contents
 * @subcategory Reducer
 */

import * as actionTypes from './actionTypes';
import { combineReducers } from 'redux';

export const INITIAL_STATE = {
  error: {},
  isLoading: {},
  contentTypes: {
    error: {},
    isLoading: {},
    result: null,
  },
  metadata: {
    error: {},
    isLoading: {},
    result: null,
  },
};

const error = (state = INITIAL_STATE.error, action) => {
  switch (action.type) {
    case actionTypes.GET_CONTENT_REQUEST:
    case actionTypes.GET_COMMERCE_PAGES_REQUEST:
    case actionTypes.GET_CONTENT_PAGES_REQUEST:
      return {
        ...state,
        [action.payload.hash]: null,
      };
    case actionTypes.GET_CONTENT_FAILURE:
    case actionTypes.GET_COMMERCE_PAGES_FAILURE:
    case actionTypes.GET_CONTENT_PAGES_FAILURE:
      return {
        ...state,
        [action.payload.hash]: action.payload.error,
      };
    default:
      return state;
  }
};

const isLoading = (
  state = INITIAL_STATE.isLoading,
  /* istanbul ignore next */ action = {},
) => {
  switch (action.type) {
    case actionTypes.GET_CONTENT_REQUEST:
    case actionTypes.GET_COMMERCE_PAGES_REQUEST:
    case actionTypes.GET_CONTENT_PAGES_REQUEST:
      return {
        ...state,
        [action.payload.hash]: true,
      };
    case actionTypes.GET_CONTENT_SUCCESS:
    case actionTypes.GET_COMMERCE_PAGES_SUCCESS:
    case actionTypes.GET_CONTENT_PAGES_SUCCESS:
      return {
        ...state,
        [action.payload.hash]: false,
      };
    case actionTypes.GET_CONTENT_FAILURE:
    case actionTypes.GET_COMMERCE_PAGES_FAILURE:
    case actionTypes.GET_CONTENT_PAGES_FAILURE:
      return {
        ...state,
        [action.payload.hash]: undefined,
      };
    default:
      return state;
  }
};

const contentTypes = (
  state = INITIAL_STATE.contentTypes,
  /* istanbul ignore next */ action = {},
) => {
  switch (action.type) {
    case actionTypes.GET_CONTENT_TYPES_REQUEST:
      return {
        isLoading: true,
        error: INITIAL_STATE.contentTypes.error,
      };
    case actionTypes.GET_CONTENT_TYPES_SUCCESS:
      return {
        result: action.payload,
        isLoading: false,
      };
    case actionTypes.GET_CONTENT_TYPES_FAILURE:
      return {
        isLoading: false,
        error: action.payload.error,
      };
    default:
      return state;
  }
};

const metadata = (
  state = INITIAL_STATE.metadata,
  /* istanbul ignore next */ action = {},
) => {
  switch (action.type) {
    case actionTypes.GET_SEO_REQUEST:
      return {
        ...state,
        isLoading: {
          ...state.isLoading,
          [action.payload.pathname]: true,
        },
        error: {
          ...state.error,
          [action.payload.pathname]: null,
        },
      };
    case actionTypes.GET_SEO_SUCCESS:
      return {
        ...state,
        result: {
          ...state.result,
          [action.payload.pathname]: action.payload.result,
        },
        isLoading: {
          ...state.isLoading,
          [action.payload.pathname]: false,
        },
      };
    case actionTypes.GET_SEO_FAILURE:
      return {
        ...state,
        isLoading: {
          ...state.isLoading,
          [action.payload.pathname]: undefined,
        },
        error: {
          ...state.error,
          [action.payload.pathname]: action.payload.error,
        },
      };
    default:
      return state;
  }
};

export const getError = state => state.error;
export const getIsLoading = state => state.isLoading;
export const getContentType = state => state.contentTypes.result;
export const getSEOmetadata = state => state.metadata;

const reducers = combineReducers({
  error,
  isLoading,
  contentTypes,
  metadata,
});

/**
 * Reducer for contents state.
 *
 * @function contentsReducer
 * @static
 *
 * @param {object} state - Current redux state.
 * @param {object} action - Action dispatched.
 *
 * @returns {object} New state.
 */
export default (state, action = {}) => {
  if (action.type === actionTypes.RESET_CONTENTS) {
    return reducers(INITIAL_STATE, action);
  }

  return reducers(state, action);
};
