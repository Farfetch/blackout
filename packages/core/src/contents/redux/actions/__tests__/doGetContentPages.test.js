import { doGetContentPages } from '..';
import {
  expectedContentPagesNormalizedPayload,
  mockContentPages,
  typeContent,
} from 'tests/__fixtures__/contents';
import { mockStore } from '../../../../../tests';
import find from 'lodash/find';
import reducer, { actionTypes } from '../..';

jest.mock('../../../utils', () => ({
  buildContentGroupHash: () => 'content_pages!woman',
  buildSEOPathname: jest.fn(),
}));

const contentPagesMockStore = (state = {}) =>
  mockStore({ contents: reducer() }, state);

const normalizeSpy = jest.spyOn(require('normalizr'), 'normalize');
const expectedConfig = undefined;
let store;

describe('doGetContentPages() action creator', () => {
  const getContentPages = jest.fn();
  const action = doGetContentPages(getContentPages);
  const slug = '/shopping/test';

  beforeEach(() => {
    jest.clearAllMocks();
    store = contentPagesMockStore();
  });

  it('should create the correct actions for when the get content pages procedure fails', async () => {
    const expectedError = new Error('Get content pages error');

    getContentPages.mockRejectedValueOnce(expectedError);

    expect.assertions(4);

    try {
      await store.dispatch(action(slug, typeContent));
    } catch (error) {
      expect(error).toBe(expectedError);
      expect(getContentPages).toHaveBeenCalledTimes(1);
      expect(getContentPages).toHaveBeenCalledWith(
        typeContent,
        { slug, strategy: undefined },
        expectedConfig,
      );
      expect(store.getActions()).toEqual(
        expect.arrayContaining([
          expect.objectContaining(
            {
              type: actionTypes.GET_CONTENT_PAGES_FAILURE,
            },
            {
              payload: { error: expectedError },
              type: actionTypes.GET_CONTENT_PAGES_FAILURE,
            },
          ),
        ]),
      );
    }
  });

  it('should create the correct actions for when the get content pages procedure is successful', async () => {
    getContentPages.mockResolvedValueOnce(mockContentPages);

    expect.assertions(5);

    await store.dispatch(action(slug, typeContent));

    const actionResults = store.getActions();

    expect(normalizeSpy).toHaveBeenCalledTimes(1);
    expect(getContentPages).toHaveBeenCalledTimes(1);
    expect(getContentPages).toHaveBeenCalledWith(
      typeContent,
      { slug, strategy: undefined },
      expectedConfig,
    );
    expect(store.getActions()).toEqual(
      expect.arrayContaining([
        expect.objectContaining(
          {
            type: actionTypes.GET_CONTENT_PAGES_REQUEST,
          },
          {
            payload: expectedContentPagesNormalizedPayload,
            type: actionTypes.GET_CONTENT_PAGES_SUCCESS,
          },
        ),
      ]),
    );
    expect(
      find(actionResults, { type: actionTypes.GET_CONTENT_PAGES_SUCCESS }),
    ).toMatchSnapshot('Get content pages payload');
  });
});
