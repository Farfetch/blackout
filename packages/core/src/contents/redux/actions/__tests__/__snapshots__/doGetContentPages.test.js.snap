// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`doGetContentPages() action creator should create the correct actions for when the get content pages procedure is successful: Get content pages payload 1`] = `
Object {
  "meta": Object {
    "query": Object {
      "slug": "/shopping/test",
      "strategy": undefined,
    },
  },
  "payload": Object {
    "entities": Object {
      "contentGroups": Object {
        "content_pages!woman": Object {
          "entries": Array [
            "dc9c0c95-9485-45c2-a76c-6923bb39b544",
          ],
          "hash": "content_pages!woman",
          "number": 1,
          "totalItems": 1,
          "totalPages": 1,
        },
      },
      "contents": Object {
        "dc9c0c95-9485-45c2-a76c-6923bb39b544": Object {
          "code": "LISTING_gender_0/brand_5030844/category_136643",
          "components": Array [
            Object {
              "customType": "article",
              "displayOptions": Object {},
              "fields": Object {
                "content": Object {
                  "content": "<p>What was once a keepsake from an island getaway has now become a permanent fixture in your jewellery collection thanks to Gigi Clozeau. Browsing her collection of anklets will keep you in the holiday mood all year round, so much so, you&#39;ll want one in every colour for every destination. Crafted from 18K gold, these delicate chains are perfectly accented with the addition of her signature resin beads in a mix of tropical colours. Browse Gigi Clozeau body jewellery below to be <em>vacay</em> ready at all times.</p>",
                  "displayOptions": Object {},
                  "name": "Listing Page Description 1",
                  "type": "html",
                },
                "title": Object {
                  "displayOptions": Object {},
                  "name": "test",
                  "type": "text",
                  "value": "Commerce Page",
                },
              },
              "name": "article",
              "type": "custom",
            },
          ],
          "contentTypeCode": "content_pages",
          "environmentCode": "live",
          "id": undefined,
          "metadata": Object {
            "custom": Object {
              "brand": "5030844",
              "category": "136643",
              "eventDate": "0001-01-01T00:00:00Z",
              "gender": "0",
              "id": "",
              "none": "",
              "priceType": "0",
              "sku": "",
              "slug": "",
              "type": "LISTING",
            },
          },
          "publicationDate": "2021-08-20T15:40:39.0000132Z",
          "publicationId": "dc9c0c95-9485-45c2-a76c-6923bb39b544",
          "spaceCode": "website",
          "target": Object {
            "contentzone": "10674",
          },
          "versionId": "78f1922d-0ef1-46ed-b02c-ca541d0a0d80",
        },
      },
    },
    "hash": "content_pages!woman",
    "result": "content_pages!woman",
  },
  "type": "@farfetch/blackout-core/GET_COMMERCE_PAGES_SUCCESS",
}
`;
